File: Java/src/PermutationsII.java
import java.util.*;

/**
 * Given a collection of numbers that might contain duplicates, return all possible unique permutations.
 * <p>
 * For example,
 * [1,1,2] have the following unique permutations:
 * [
 * [1,1,2],
 * [1,2,1],
 * [2,1,1]
 * ]
 * <p>
 * Accepted.
 */
public class PermutationsII {

    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> results = new LinkedList<>();

        if (nums.length == 0) return results;
        if (nums.length == 1) {
            List<Integer> list = new LinkedList<>();
            list.add(nums[0]);
            results.add(list);
            return results;
        }

        int[] ints = new int[nums.length - 1];
        System.arraycopy(nums, 0, ints, 0, nums.length - 1);

        Set<List<Integer>> set = new HashSet<>();

        for (List<Integer> list : permuteUnique(ints)) {
            for (int i = 0; i <= list.size(); i++) {
                List<Integer> tmp = new LinkedList<>(list);
                tmp.add(i, nums[nums.length - 1]);
                set.add(tmp);
            }
        }

        results.addAll(set);
        return results;
    }

}

File: Kotlin/src/PermutationsII.kt
/**
 * Given a collection of numbers that might contain duplicates, return all possible unique permutations.
 *
 * For example,
 * [1,1,2] have the following unique permutations:
 * [
 * [1,1,2],
 * [1,2,1],
 * [2,1,1]
 * ]
 *
 * Accepted.
 */
class PermutationsII {

    fun permuteUnique(nums: IntArray): List<List<Int>> {
        val results = mutableListOf<List<Int>>()

        if (nums.isEmpty()) return results
        if (nums.size == 1) return results.apply { add(mutableListOf(nums[0])) }

        val ints = IntArray(nums.size - 1)
        System.arraycopy(nums, 0, ints, 0, nums.size - 1)

        val set = mutableSetOf<List<Int>>()

        for (list in permuteUnique(ints)) {
            for (i in 0..list.size) {
                val tmp = mutableListOf<Int>()
                tmp.addAll(list)
                tmp.add(i, nums[nums.size - 1])
                set.add(tmp)
            }
        }

        return results.apply { addAll(set) }
    }

    companion object {

        @JvmStatic
        fun main(args: Array<String>) {
            val p = PermutationsII()

            // Expected: [[]]
            println(p.permuteUnique(intArrayOf()))

            // Expected: [[1]]
            println(p.permuteUnique(intArrayOf(1)))

            // Expected: [[1, 1, 2], [1, 2, 1], [2, 1, 1]]
            println(p.permuteUnique(intArrayOf(1, 1, 2)))
        }
    }

}
